### DISCUSSION TITLE
[v9] How to create network requests spies and mocks with dynamic responses?

### DISCUSSION THREAD

**Discussion by ls-bit** [webdriverio/webdriverio#13729] [Q&A] [answered]
Hello,

i am trying to setup some tests with network request mocks and spies, and cannot get it to work how it is written in documentation.

## Request spy
According to https://webdriver.io/docs/mocksandspies#spies the `mock.respond` call should not be needed, but without it does not work. 

Example:
```typescript
    it('request spy', async () => {
        const mock = await browser.mock('https://the-internet.herokuapp.com/js/foundation/foundation.js', {
            method: 'get',
            statusCode: 200,
        });
        // uncomment following for test to pass, but it modifies the response body
        // mock.respond({})

        await browser.url('https://the-internet.herokuapp.com/');

        await mock.waitForResponse();
        await expect(mock.calls.length).toBe(1);
    });
```
How can I make the mock act like a spy and let it continue with the original response it would normally receive?

Edit: I think i found a possible workaround by using e.g. `mock.respond(null as any)`, which seems to result in original response being returned.

## Dynamic response
According to https://webdriver.io/docs/mocksandspies#dynamic-responses `mock.response` can take a function to provide a dynamic response. This seems to be no longer the case, the types do not include a function parameter. Is there another way to provide dynamic responses?

Are docs not up to date or am i doing something wrong? Thanks for helping

**Comment by christian-bromann** [comment, marked as answer]
Thank you for raising this. I've created a PR in #13731 with a fix.
